### PR TITLE
ARROW-17021: [C++][R][CI] Enable use of sccache in crossbow

### PR DESCRIPTION
**Pull request by assignUser**
This is a first stab at enabling sccache, I used the R nightly job as an example as it compiles arrow on mac, win and linux.
I want to open this for feedback specifically with an eye towards the cmake changes I implemented which enable this as long as the correct envvars (AWS creds + bucket) are set and sccache is available. My hope was to keep the changes required to activate sccache for existing jobs as minimal as possible.

While working on this I also noticed that `ThirdpartyToolchain.cmake` is not build using (s)ccache, the flags had to be explicitly set. That fix alone should speed up CI across the board (in arrow and crossbow). 
For the R task this results in a build time reduction of 40-70% 🚀 

Here a run of the same task without caching https://github.com/ursacomputing/crossbow/actions/runs/2563558583

cc: @kszucs @raulcd @kou

Thanks for your input!

### CODE REVIEW

**Comment by assignUser** [comment]
@github-actions crossbow submit r-binary-packages

**Comment by assignUser** [comment]
@github-actions crossbow submit r-binary-packages

**Comment by assignUser** [comment]
@github-actions crossbow submit r-binary-packages

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -752,6 +752,14 @@ set(EP_COMMON_CMAKE_ARGS
     -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=${CMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY}
     -DCMAKE_VERBOSE_MAKEFILE=${CMAKE_VERBOSE_MAKEFILE})
 
+# Enable s/ccache if set by parent.
+if(CMAKE_CXX_COMPILER_LAUNCHER AND CMAKE_CXX_COMPILER_LAUNCHER)
+  set(EP_COMMON_CMAKE_ARGS
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:757]: Could you use `list(APPEND EP_COMMON_CMAKE_ARGS ...)`?

File: dev/tasks/r/github.packages.yml [hunk 2]
```diff
@@ -114,17 +124,38 @@ jobs:
...
+          echo "SCCACHE_PATH=$(pwd)/sccache/sccache.exe" >> $GITHUB_ENV
       - name: Build Arrow C++ with rtools40
         shell: bash
         env:
           ARROW_HOME: "arrow"
+          AWS_SECRET_ACCESS_KEY: {{ '${{ secrets.AWS_SECRET_ACCESS_KEY }}' }}
+          AWS_ACCESS_KEY_ID: {{ '${{ secrets.AWS_ACCESS_KEY_ID }}' }}
+          SCCACHE_BUCKET: {{ '${{ secrets.SCCACHE_BUCKET }}' }}
+          SCCACHE_S3_KEY_PREFIX: 'sccache'
+          ARROW_USE_SCCACHE: ON
```
Comment (kou) [inline dev/tasks/r/github.packages.yml:158]: Could you sort this list?

File: dev/tasks/r/github.packages.yml [hunk 3]
```diff
@@ -82,14 +82,24 @@ jobs:
...
+
       - name: Build libarrow
         shell: bash
         env:
-          UBUNTU: {{ '${{ matrix.config.version}}' }}
+          UBUNTU: {{ '${{ matrix.config.version }}' }}
+          SC_KEY: {{' ${{ secrets.AWS_SECRET_ACCESS_KEY }}' }}
+          SC_ID: {{ '${{ secrets.AWS_ACCESS_KEY_ID }}' }}
+          SC_BUCKET: {{ '${{ secrets.SCCACHE_BUCKET }}' }}
+          SC_PREFIX: sccache
```
Comment (kou) [inline dev/tasks/r/github.packages.yml:93]: How about using `SCCACHE_` as prefix like others?

File: dev/tasks/r/github.packages.yml [hunk 4]
```diff
@@ -280,10 +319,33 @@ jobs:
...
+          curl -L $SCCACHE_URL.sha256 --output sccache.tar.gz.sha256
+
+          SCCACHE_ARCHIVE=sccache.tar.gz
+          echo "$(cat $SCCACHE_ARCHIVE.sha256) $SCCACHE_ARCHIVE" | sha256sum --check --status
+
+          mkdir -p sccache
+          tar -xzvf $SCCACHE_ARCHIVE --strip-component=1 --directory sccache
+
+          echo "$(pwd)/sccache" >> $GITHUB_PATH
+          echo "SCCACHE_PATH=$(pwd)/sccache/sccache.exe" >> $GITHUB_ENV
```
Comment (kou) [inline dev/tasks/r/github.packages.yml:339]:
```suggestion
          echo "SCCACHE_PATH=$(pwd)/sccache/sccache" >> $GITHUB_ENV

```

File: r/inst/build_arrow_static.sh [hunk 5]
```diff
@@ -81,10 +96,14 @@ ${CMAKE} -DARROW_BOOST_USE_SHARED=OFF \
     -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON \
     -DCMAKE_UNITY_BUILD=${CMAKE_UNITY_BUILD:-OFF} \
     -Dxsimd_SOURCE=${xsimd_SOURCE:-} \
+    -DARROW_USE_SCCACHE=ON \
```
Comment (kou) [inline r/inst/build_arrow_static.sh:99]: Could you keep alphabetical order?

File: dev/tasks/r/github.packages.yml [hunk 6]
```diff
@@ -280,10 +319,33 @@ jobs:
...
+
+          # Download archive and checksum
+          curl -L $SCCACHE_URL --output sccache.tar.gz
+          curl -L $SCCACHE_URL.sha256 --output sccache.tar.gz.sha256
+
+          SCCACHE_ARCHIVE=sccache.tar.gz
+          echo "$(cat $SCCACHE_ARCHIVE.sha256) $SCCACHE_ARCHIVE" | sha256sum --check --status
+
+          mkdir -p sccache
+          tar -xzvf $SCCACHE_ARCHIVE --strip-component=1 --directory sccache
```
Comment (kou) [inline dev/tasks/r/github.packages.yml:336]: Do we need `chmod +x sscache/sscache`?

File: ci/scripts/PKGBUILD [hunk 7]
```diff
@@ -128,7 +129,7 @@ build() {
     -DCMAKE_BUILD_TYPE="release" \
     -DCMAKE_INSTALL_PREFIX=${MINGW_PREFIX} \
     -DCMAKE_UNITY_BUILD=${CMAKE_UNITY_BUILD} \
-    -DCMAKE_VERBOSE_MAKEFILE=ON
+    -DCMAKE_VERBOSE_MAKEFILE=ON
```
Comment (kou) [inline ci/scripts/PKGBUILD:132]:
```suggestion
    -DCMAKE_VERBOSE_MAKEFILE=ON
```

File: cpp/CMakeLists.txt [hunk 8]
```diff
@@ -215,18 +215,39 @@ else()
   set(PYTHON_EXECUTABLE ${Python3_EXECUTABLE})
 endif()
 
-if(ARROW_USE_CCACHE
-   AND NOT CMAKE_C_COMPILER_LAUNCHER
-   AND NOT CMAKE_CXX_COMPILER_LAUNCHER)
+if((ARROW_USE_SCCACHE)
+  AND NOT CMAKE_C_COMPILER_LAUNCHER
+  AND NOT CMAKE_CXX_COMPILER_LAUNCHER)
```
Comment (kou) [inline cpp/CMakeLists.txt:220]:
```suggestion
   AND NOT CMAKE_C_COMPILER_LAUNCHER
   AND NOT CMAKE_CXX_COMPILER_LAUNCHER)
```

File: cpp/CMakeLists.txt [hunk 9]
```diff
@@ -215,18 +215,39 @@ else()
...
+
+  find_program(SCCACHE_FOUND sccache)
+  
+  if(NOT SCCACHE_FOUND AND DEFINED ENV{SCCACHE_PATH})
+    # cmake has problems finding sccache from within mingw
```
Comment (kou) [inline cpp/CMakeLists.txt:225]: Can we remove this by adding `chmod +x sccache/sccache.exe`?

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 10]
```diff
@@ -752,6 +752,14 @@ set(EP_COMMON_CMAKE_ARGS
...
+      ${EP_COMMON_CMAKE_ARGS}
+      -DCMAKE_C_COMPILER_LAUNCHER=${CMAKE_C_COMPILER_LAUNCHER}
+      -DCMAKE_CXX_COMPILER_LAUNCHER=${CMAKE_CXX_COMPILER_LAUNCHER})
+endif()
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:761]: Wow! Can we split this as a separated pull request to merge only this without sscache change?

File: r/inst/build_arrow_static.sh [hunk 11]
```diff
@@ -81,10 +96,14 @@ ${CMAKE} -DARROW_BOOST_USE_SHARED=OFF \
...
     ${EXTRA_CMAKE_FLAGS} \
     -G ${CMAKE_GENERATOR:-"Unix Makefiles"} \
     ${SOURCE_DIR}
 
 ${CMAKE} --build . --target install
 
+echo "=== sccache stats after the build ==="
+sccache --show-stats
```
Comment (kou) [inline r/inst/build_arrow_static.sh:107]: Could you show a build log that we can see this?

Comment (assignUser) [reply r/inst/build_arrow_static.sh:107]: https://github.com/ursacomputing/crossbow/runs/7257889831?check_suite_focus=true#step:6:2578

The other Linux builds of that matrix and the wincpp job also displays them if you want to compare :D

File: cpp/cmake_modules/DefineOptions.cmake [hunk 12]
```diff
@@ -111,7 +111,10 @@ if(ARROW_DEFINE_OPTIONS)
 
   define_option(ARROW_POSITION_INDEPENDENT_CODE
                 "Whether to create position-independent target" ON)
-
+  
+  define_option(ARROW_USE_SCCACHE "Use sccache when compiling (if available), takes precdent\
```
Comment (raulcd) [inline cpp/cmake_modules/DefineOptions.cmake:115]:
```suggestion
  define_option(ARROW_USE_SCCACHE "Use sccache when compiling (if available), takes precedent\
```

File: dev/tasks/r/github.packages.yml [hunk 13]
```diff
@@ -155,11 +186,19 @@ jobs:
           rtools-version: {{ '${{ matrix.r_version.rtools }}' }}
           Ncpus: 2
       {{ macros.github_setup_local_r_repo(false, true)|indent }}
+      - name: Install sccache
+        if: startsWith(matrix.platform, 'macos')
+        run: brew install sccache
```
Comment (raulcd) [inline dev/tasks/r/github.packages.yml:191]: could we install on windows following the same approach we don on line 132: `- name: Get sccache`?
Maybe we could create our own action for that. I am just thinking out loud, not required for this PR.

File: dev/tasks/r/github.packages.yml [hunk 14]
```diff
@@ -280,10 +319,33 @@ jobs:
...
+
+          SCCACHE_ARCHIVE=sccache.tar.gz
+          echo "$(cat $SCCACHE_ARCHIVE.sha256) $SCCACHE_ARCHIVE" | sha256sum --check --status
+
+          mkdir -p sccache
+          tar -xzvf $SCCACHE_ARCHIVE --strip-component=1 --directory sccache
+
+          echo "$(pwd)/sccache" >> $GITHUB_PATH
+          echo "SCCACHE_PATH=$(pwd)/sccache/sccache.exe" >> $GITHUB_ENV
+      - run: sudo apt update && sudo apt install libcurl4-openssl-dev
```
Comment (raulcd) [inline dev/tasks/r/github.packages.yml:340]: why is this now required? Should be done on the previous step (install sscache)?

**Comment by kou** [comment]
We want to use `sccache` because we can't use `actions/cache` of GitHub Actions with our "one branch per task" approach, right?

Can we change our approach so that we can use `actions/cache`? For example:

1. We always use the `master` branch for all tasks instead of creating one branch per task
2. We add `jobs/${JOB_ID}/tasks.yaml` (or something) that lists commit IDs for target tasks instead of using branch search feature of GitHub (that is too slow for many branches)
    (We can keep only the latest N jobs because old logs in GitHub Actions are removed.)
3. We create a `${JOB_ID}-${TASK_NAME}` tag for each commit of task to create a release of GitHub Releases to upload artifacts

**Comment by pitrou** [comment]
> Can we change our approach so that we can use `actions/cache`?

We can, but we'll also run into capacity limits which may make caching poorly efficient ("A repository can have up to 10GB of caches").

File: cpp/cmake_modules/DefineOptions.cmake (hunk 12 above)
Comment (pitrou) [inline cpp/cmake_modules/DefineOptions.cmake:115]:
```suggestion
  define_option(ARROW_USE_SCCACHE "Use sccache when compiling (if available), takes precdence\
```

File: dev/tasks/r/github.packages.yml [hunk 15]
```diff
@@ -114,17 +124,38 @@ jobs:
...
       - uses: r-lib/actions/setup-r@v2
         with:
           rtools-version: 40
           r-version: "4.0"
           Ncpus: 2
+      - name: Get sccache
+        shell: bash
+        run: |
+          SCCACHE_BUILD=pc-windows-msvc
+          SCCACHE_URL="https://github.com/mozilla/sccache/releases/download/v0.3.0/sccache-v0.3.0-x86_64-$SCCACHE_BUILD.tar.gz"
```
Comment (pitrou) [inline dev/tasks/r/github.packages.yml:136]: Should there be a script e.g. `ci/script/install_sccache.sh` for this?

The syntax could be `ci/scripts/install_sccache.sh <build> <executable install dir>`

File: dev/tasks/r/github.packages.yml (hunk 14 above)
Comment (pitrou) [reply dev/tasks/r/github.packages.yml:340]: Seems a bit weird indeed. Why would dev dependencies be neede for a compiled executable?

File: dev/tasks/r/github.packages.yml (hunk 13 above)
Comment (assignUser) [reply dev/tasks/r/github.packages.yml:191]: Yes I was also already thinking of making this it's own action.

**Comment by assignUser** [comment]
@github-actions crossbow submit r-binary-packages

File: cpp/CMakeLists.txt (hunk 9 above)
Comment (assignUser) [reply cpp/CMakeLists.txt:225]: No it falls back to [ccache](https://github.com/ursacomputing/crossbow/runs/7388502838?check_suite_focus=true#step:7:5555) if we don't use the envvar hack. I would prefer to not need it too...

**Comment by assignUser** [comment]
@kou adding to @pitrou's comment: We could rework archery to take advantage of `actions/cache` but the action itself is inefficient/broken for this use case because it does not update on a primary key hit, this would force us to create new, unique caches every time leading to the mentioned capacity issue. 

It would also be a **major** rework of all of the related archery functions, which is likely to happen at some point but sccache will help us *now*.

File: dev/tasks/r/github.packages.yml (hunk 14 above)
Comment (assignUser) [reply dev/tasks/r/github.packages.yml:341]: This is a source build and does not use the binaries, so these are required for aws/gcs bindings to be enabled.

**Comment by kszucs** [comment]
> For all the reasons above I think we shouldn't bother with `actions/cache` here.

Agree.

File: dev/tasks/r/github.packages.yml [hunk 16]
```diff
@@ -82,14 +82,24 @@ jobs:
...
       {{ macros.github_change_r_pkg_version(is_fork, '${{ needs.source.outputs.pkg_version }}')|indent }}
       {{ macros.github_install_archery()|indent }}
+
       - name: Build libarrow
         shell: bash
         env:
-          UBUNTU: {{ '${{ matrix.config.version}}' }}
+          SCCACHE_BUCKET: {{ '${{ secrets.SCCACHE_BUCKET }}' }}
+          SCCACHE_ID: {{ '${{ secrets.AWS_ACCESS_KEY_ID }}' }}
+          SCCACHE_KEY: {{' ${{ secrets.AWS_SECRET_ACCESS_KEY }}' }}
```
Comment (raulcd) [inline dev/tasks/r/github.packages.yml:91]: wouldn't be better to call this `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`. I am not sure the renaming adds anything and the environment variable names used on archery are already the original ones.

**Review by raulcd — COMMENTED** [review]
Thanks for this @assignUser!

File: dev/tasks/r/github.packages.yml (hunk 16 above)
Comment (assignUser) [reply dev/tasks/r/github.packages.yml:91]: The idea was to make the archery command a bit more concise as archery emulates the docker `-e` param but does not support passing only the envvar name.

**Comment by assignUser** [comment]
@github-actions crossbow submit r-binary-packages

**Comment by assignUser** [comment]
@github-actions crossbow submit r-binary-packages

File: dev/tasks/r/github.packages.yml (hunk 16 above)
Comment (pitrou) [reply dev/tasks/r/github.packages.yml:91]: Also, I like that this is explicit that these are for sccache, not for unrelated purposes. So I would vote to keep the `SCCACHE` prefix.

File: dev/tasks/r/github.packages.yml [hunk 17]
```diff
@@ -82,14 +82,24 @@ jobs:
...
+          SCCACHE_PREFIX: sccache
```
Comment (pitrou) [inline dev/tasks/r/github.packages.yml:92]: SCCACHE_KEY_PREFIX perhaps?

Comment (raulcd) [reply dev/tasks/r/github.packages.yml:92]: I am happier with this name `SCCACHE_S3_KEY_PREFIX` too as it gives me a little more context.

File: cpp/cmake_modules/DefineOptions.cmake [hunk 18]
```diff
@@ -111,7 +111,10 @@ if(ARROW_DEFINE_OPTIONS)
 
   define_option(ARROW_POSITION_INDEPENDENT_CODE
                 "Whether to create position-independent target" ON)
-
+  
+  define_option(ARROW_USE_SCCACHE "Use sccache when compiling (if available), takes precdence\
+   over ccache" ON)
```
[the remark this answers is not in the document]
Comment (assignUser) [reply cpp/cmake_modules/DefineOptions.cmake:116]: If there are no envvars set sccache will function in local mode similar to ccache https://github.com/mozilla/sccache#local so I would tend to leave it on as it allows us to just add the install script to the docker-compose command and the envvars to the github template to enable this on a lot of jobs, We could of course also switch it on in the build scripts but we would need to touch a bunch more files. 

Alternatively we could check if a storage back end is configured and only use sccache over ccache in that case if both are installed.

**Comment by assignUser** [comment]
@github-actions crossbow submit r-binary-packages conan-*

**Comment by assignUser** [comment]
@github-actions crossbow submit r-binary-packages conan-*

**Comment by assignUser** [comment]
@github-actions crossbow submit r-binary-packages conan-*

**Comment by assignUser** [comment]
@github-actions crossbow submit conan-*

**Comment by kou** [comment]
> It would also be a **major** rework of all of the related archery functions, which is likely to happen at some point but sccache will help us _now_.

OK.

**Comment by assignUser** [comment]
@pitrou see above, it is now very easy to activate sccache for tasks
I re-ran `test-conda-cpp` 32min -> 13min

File: dev/tasks/r/github.packages.yml [hunk 19]
```diff
@@ -82,10 +82,15 @@ jobs:
...
       {{ macros.github_install_archery()|indent }}
+
       - name: Build libarrow
         shell: bash
         env:
-          UBUNTU: {{ '${{ matrix.config.version}}' }}
+          AWS_ACCESS_KEY_ID: {{ '${{ secrets.AWS_ACCESS_KEY_ID }}' }}
+          AWS_SECRET_ACCESS_KEY: {{' ${{ secrets.AWS_SECRET_ACCESS_KEY }}' }}
+          SCCACHE_BUCKET: {{ '${{ secrets.SCCACHE_BUCKET }}' }}
+          SCCACHE_S3_KEY_PREFIX: sccache
```
Comment (pitrou) [inline dev/tasks/r/github.packages.yml:92]: This snippet is everywhere, can it be made a macro?

File: docker-compose.yml [hunk 20]
```diff
@@ -62,6 +62,12 @@ x-ccache: &ccache
   CCACHE_MAXSIZE: 1G
   CCACHE_DIR: /ccache
 
+x-sccache: &sccache
+  AWS_ACCESS_KEY_ID:
+  AWS_SECRET_ACCESS_KEY:
+  SCCACHE_BUCKET:
```
Comment (pitrou) [inline docker-compose.yml:68]: Does it mean they are inherited from the calling process environment?

File: docker-compose.yml [hunk 21]
```diff
@@ -245,8 +251,10 @@ services:
     volumes: &conda-volumes
       - .:/arrow:delegated
       - ${DOCKER_VOLUME_PREFIX}conda-ccache:/ccache:delegated
-    command:
-      ["/arrow/ci/scripts/cpp_build.sh /arrow /build &&
+    command: &conda-cpp-command
+      ["/arrow/ci/scripts/install_sccache.sh unknown-linux-musl &&
```
Comment (pitrou) [inline docker-compose.yml:255]: I'm curious: why "unknown-linux-musl"? Does our conda image really use musl?

File: ci/scripts/install_sccache.sh [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+curl -L $SCCACHE_URL --output $SCCACHE_ARCHIVE
+curl -L $SCCACHE_URL.sha256 --output $SCCACHE_ARCHIVE.sha256
+
+echo "$(cat $SCCACHE_ARCHIVE.sha256) $SCCACHE_ARCHIVE" | sha256sum --check --status
+
+mkdir -p sccache
+tar -xzvf $SCCACHE_ARCHIVE --strip-component=1 --directory sccache
+chmod u+x sccache/sccache
+
+SCCACHE_DIR=$(pwd)/sccache
```
Comment (pitrou) [inline ci/scripts/install_sccache.sh:45]: Can the install prefix instead be taken as an argument, just like for `install_minio.sh`?

File: docker-compose.yml (hunk 21 above)
Comment (pitrou) [reply docker-compose.yml:255]: Also, can the `install_sccache.sh` call instead be moved to the dockerfile?

File: docker-compose.yml [hunk 23]
```diff
@@ -245,8 +251,10 @@ services:
...
+        export PATH=$$(pwd)/sccache:$$PATH &&
```
Comment (pitrou) [inline docker-compose.yml:256]: Probably sccache can be installed to `/usr/local/bin` instead?

File: r/inst/build_arrow_static.sh [hunk 24]
```diff
@@ -87,4 +87,7 @@ ${CMAKE} -DARROW_BOOST_USE_SHARED=OFF \
...
```
Comment (pitrou) [inline r/inst/build_arrow_static.sh:91]: Can we add similar snippets to `ci/script/cpp_build.sh`, just as with ccache?

**Review by pitrou — COMMENTED** [review]
Thanks a lot, this seems very promising.

**Comment by pitrou** [comment]
Caveat: looking at the source code, sccache only supports setting a cache size limit for local disk and for Redis. So our S3 bucket size may grow a lot if we are not careful.

File: docker-compose.yml (hunk 21 above)
Comment (assignUser) [reply docker-compose.yml:255]: > There are prebuilt x86-64 binaries available for Windows, **Linux (a portable binary compiled against musl),** and macOS 

I haven't moved the installation to the dockerfiles for ease of testing but we can do that.

File: docker-compose.yml (hunk 20 above)
Comment (assignUser) [reply docker-compose.yml:68]: Yes, saving us the work to add the mall manually via `-e`

**Comment by assignUser** [comment]
@github-actions crossbow submit conan-* test-conda-cpp test-conda-python-* test-conda-python-3.8-hypothesis r-binary-packages

File: ci/scripts/PKGBUILD [hunk 25]
```diff
@@ -73,6 +73,9 @@ build() {
   # set the appropriate compiler definition.
   export CPPFLAGS="-DUTF8PROC_STATIC"
 
+  # CMAKE_UNITY_BUILD is set to OFF as otherwise some compute functionality
```
Comment (assignUser) [inline ci/scripts/PKGBUILD:76]: Added this comment to avoid a minor follow up PR from @paleolimbot

**Comment by assignUser** [comment]
crossbow submit conan-* test-conda-cpp test-conda-python-*

**Comment by assignUser** [comment]
@github-actions crossbow submit conan-* test-conda-cpp test-conda-python-*

**Comment by assignUser** [comment]
The failures above (except conan) are all tests failures, sccache works and is used: 
image

**Comment by assignUser** [comment]
@github-actions crossbow submit conan-* r-binary-packages

**Comment by assignUser** [comment]
@github-actions crossbow submit conan-*

**Comment by assignUser** [comment]
@github-actions crossbow submit r-binary-packages

**Comment by assignUser** [comment]
This is ready for review, the job above only failed due to the missing [centos artifact, ](https://github.com/ursacomputing/crossbow/actions/runs/3052342834/jobs/4924218048#step:10:37)I have fixed that in `task.yml`.

**Comment by assignUser** [comment]
Sorry for the review request spam, the github frontend seems to be broken for me at the moment. @kou @pitrou could you have a look?

File: ci/scripts/install_sccache.sh [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+    echo "Will default to arch=x86_64 and version=0.3.0 "
+    exit 1
+fi
+
+BUILD=$1
+PREFIX=$2
+ARCH=${3:-x86_64}
+VERSION=${4:-0.3.0}
+
+SCCACHE_URL="https://github.com/mozilla/sccache/releases/download/v0.3.0/sccache-v$VERSION-$ARCH-$BUILD.tar.gz"
```
Comment (pitrou) [inline ci/scripts/install_sccache.sh:33]: Did you mean
```suggestion
SCCACHE_URL="https://github.com/mozilla/sccache/releases/download/v$VERSION/sccache-v$VERSION-$ARCH-$BUILD.tar.gz"
```

File: ci/scripts/install_sccache.sh [hunk 27]
```diff
@@ -0,0 +1,53 @@
...
+SCCACHE_ARCHIVE=sccache.tar.gz
+
+# Download archive and checksum
+curl -L $SCCACHE_URL --output $SCCACHE_ARCHIVE
+curl -L $SCCACHE_URL.sha256 --output $SCCACHE_ARCHIVE.sha256
```
Comment (pitrou) [inline ci/scripts/install_sccache.sh:38]: Just for the record, why not use `wget`?

File: cpp/cmake_modules/DefineOptions.cmake [hunk 28]
```diff
@@ -111,7 +111,10 @@ if(ARROW_DEFINE_OPTIONS)
 
   define_option(ARROW_POSITION_INDEPENDENT_CODE
                 "Whether to create position-independent target" ON)
-
+  
+  define_option(ARROW_USE_SCCACHE "Use sccache when compiling (if available),;\
+takes precdence over ccache if a storage backend is configured" ON)
```
Comment (pitrou) [inline cpp/cmake_modules/DefineOptions.cmake:116]:
```suggestion
  define_option(ARROW_USE_SCCACHE "Use sccache when compiling (if available), \
takes precedence over ccache if a storage backend is configured" ON)
```

Comment (pitrou) [reply cpp/cmake_modules/DefineOptions.cmake:116]: Also, move this below ARROW_USE_CCACHE?

File: dev/tasks/tasks.yml [hunk 29]
```diff
@@ -951,7 +951,7 @@ tasks:
       custom_version: Unset
     artifacts:
       - r-lib__libarrow__bin__windows__arrow-[0-9\.]+\.zip
-      - r-lib__libarrow__bin__centos-7__arrow-[0-9\.]+\.zip
+      #- r-lib__libarrow__bin__centos-7__arrow-[0-9\.]+\.zip
```
Comment (pitrou) [inline dev/tasks/tasks.yml:954]: Add a comment explaining why?

File: docker-compose.yml [hunk 30]
```diff
@@ -411,14 +416,16 @@ services:
...
-      <<: *ccache
+      <<: [*ccache, *sccache]
       ARROW_HOME: /arrow
       ARROW_DEPENDENCY_SOURCE: BUNDLED
       LIBARROW_MINIMAL: "false"
       ARROW_MIMALLOC: "ON"
     volumes: *ubuntu-volumes
-    command: /bin/bash -c "
-        cd /arrow && r/inst/build_arrow_static.sh"
+    command: &arrow-static-cmd
```
Comment (kou) [inline docker-compose.yml:425]: Could you use the same naming convention others use? 

```suggestion
    command: &cpp-static-command
```

**Review by kou — APPROVED** [review]
+1

File: ci/scripts/install_sccache.sh (hunk 27 above)
Comment (assignUser) [reply ci/scripts/install_sccache.sh:38]: I initially used wget but that was missing from some runner (windows maybe? it's been a while ^^) so I switch to curl, only to then find out that curl is not installed in all images so  wither would have been fine :shrug:

File: ci/scripts/install_sccache.sh (hunk 26 above)
Comment (assignUser) [reply ci/scripts/install_sccache.sh:33]: Indeed, thanks

File: ci/scripts/install_sccache.sh [hunk 31]
```diff
@@ -0,0 +1,53 @@
...
+echo "$(cat $SCCACHE_ARCHIVE.sha256) $SCCACHE_ARCHIVE" | sha256sum --check --status
+
+if [ ! -d $PREFIX ]; then
+    mkdir -p $PREFIX
+fi
+
+tar -xzvf $SCCACHE_ARCHIVE --strip-component=1 --directory $PREFIX --wildcards sccache*/sccache* 
+chmod u+x $PREFIX/sccache
+
+if [ "$GITHUB_ACTIONS" ]; then
```
[the remark this answers is not in the document]
Comment (assignUser) [reply ci/scripts/install_sccache.sh:49]: I use GITHUB_ACTION***S*** which is exactly for this use case:
> GITHUB_ACTIONS | Always set to true when GitHub Actions is running the workflow. You can use this variable to differentiate when tests are being run locally or by GitHub Actions.

**Comment by pitrou** [comment]
By the way, compiling sccache from source only happened once, I hope (not on every CI run)?

File: dev/tasks/docker-tests/github.linux.yml [hunk 32]
```diff
@@ -31,6 +31,8 @@ jobs:
 
       - name: Execute Docker Build
         shell: bash
+        env:
+        {{ macros.github_set_sccache_envvars()|indent(8) }}
```
[the remark this answers is not in the document]
Comment (assignUser) [reply dev/tasks/docker-tests/github.linux.yml:35]: Indentation with jinja is a bit hard to determine wihtout actually rendering the template in my experience. There is some trimming happening by default and than this is a nested macro that adds another layer of  whitespace, I tried stripping manually but that also removes the new line after `env:`  and messes everything up completely.... 

This template is used e.g. in the conan builds and renders correctly with this setting:
```yml      
     - name: Execute Docker Build
        shell: bash
        env:
          AWS_SECRET_ACCESS_KEY: "${{ secrets.AWS_SECRET_ACCESS_KEY }}"
          AWS_ACCESS_KEY_ID: "${{ secrets.AWS_ACCESS_KEY_ID }}"
          SCCACHE_BUCKET: "${{ secrets.SCCACHE_BUCKET }}"
          SCCACHE_S3_KEY_PREFIX: "sccache"
```

File: cpp/cmake_modules/DefineOptions.cmake (hunk 28 above)
Comment (assignUser) [reply cpp/cmake_modules/DefineOptions.cmake:116]: I have moved ccache above this, the semicolon is not a typo though, it is required as a line break:
```
CMake Error at cmake_modules/DefineOptions.cmake:27 (message):
  description for ARROW_USE_SCCACHE contained a

          line 105 characters long!
          (max is 80). Split it into more lines with semicolons
```

**Comment by assignUser** [comment]
> By the way, compiling sccache from source only happened once, I hope (not on every CI run)?

That shouldn't have happened at al 🤦  it was very specific to the use of the 10.13 runner which is eol -> no bottles.  But yes it will only happen this one time, anything else would be crazy :D Should we add further 10.13 runners I will install it manually before the runner is connected.

**Comment by assignUser** [comment]
Thanks for the reviews @kou @pitrou
I have addressed your comments either via code or comments.

**Comment by assignUser** [comment]
There is a lint error in R for which there already is a PR. I have fixed the cmake issues wit harchery.

File: ci/scripts/install_sccache.sh (hunk 31 above)
[the remark this answers is not in the document]
Comment (kou) [reply ci/scripts/install_sccache.sh:49]: Ah, sorry. I missed "S".
Then, `[ "${GITHUB_ACTIONS}" = "true" ]` is better.

[the remark this answers is not in the document]
Comment (assignUser) [reply ci/scripts/install_sccache.sh:49]: Hm I guess people could set the envvar to false for their own scripts or something and that would be a false positive like this, ok 👍

**Review by kou — APPROVED** [review]
+1

File: ci/scripts/cpp_build.sh [hunk 33]
```diff
@@ -181,6 +181,11 @@ if [ "${ARROW_USE_CCACHE}" == "ON" ]; then
     ccache -s
 fi
 
+if command -v sccache &> /dev/null; then
+  echo "=== sccache stats after the build ==="
+  sccache --show-stats
```
Comment (kou) [inline ci/scripts/cpp_build.sh:186]: @assignUser This causes an error on non-amd64 environment: https://app.travis-ci.com/github/apache/arrow/jobs/583028972#L3378

```text
+ sccache --show-stats
/arrow/ci/scripts/cpp_build.sh: line 186: /usr/local/bin/sccache: cannot execute binary file: Exec format error
ERROR: 126
```

Could you fix this as a follow-up task?

**Pull request merged**